### DISCUSSION TITLE
Mass Paste Alpha from and to Multiple Files?

### DISCUSSION THREAD

**Discussion by Surronti** [ImageMagick/ImageMagick#5800] [Help]
Hi, I hope I'm writing in the right place. I'm a complete beginner with ImageMagick.

Here's the thing: I have generated a large number of normal, roughness, and displacement maps within a folder structure (5000 for each type). So for example within one folder you have alit_n.png, alit_displacement.png, and alit_rougness.png, and a bunch more triplets of this sort.

I'd like to add the inverted roughness maps as alpha channel to their normal maps, while preserving the folder structure. So pick alit_roughness, invert it, and paste it as the alpha of alit_n.

When it's done, alit_n should also have been converted to DDS.

Is there a way to batch do this with ImageMagick? 

I'm using 7.1.0-52 Q16-HDRI x64 on Windows 10.

I used to do it manually with GIMP, by using windows search to show all _n and _roughness files in the subfolders, then opening two GIMP windows and dragging all the normals into one and all roughness maps into the other and then using macros to jump between the two. GIMP however doesn't let me drag and drop hundreds of files at a time anymore :( and it was a tedious process anyway, so I'm looking for an alternative.

I can add a link to the archive I'm working on if anyone needs to take a look.

**Comment by snibgo** [comment]
What version of IM, on what platform?

IM doesn't have powerful batch facilities. It is fine for some simple tasks, but is limited. In general, you may need to write a shell script.

So you need to know (a) what IM command does what you need, and (b) how to loop through the triplets in your directory.

For (a), the IM comand might be something like:
```
magick alit_n.png ( alit_rougness.png -alpha extract -negate ) -alpha off -compose CopyOpacity -composite newdir/alit_n.png
```

For (b), I suppose the first part of the filename determines the triplet. Is that right? Then you need to do some string-handling in your shell script.

**Reply by Surronti** [reply]
Yes, you are correct: the first part of the name determines the triplet. Thanks for the answer, it's 7.1.0-52 Q16-HDRI x64 on Windows 10. Do you know if there is a resource for string-handling?

**Comment by snibgo** [comment]
In Windows BAT, I would have a "for" loop to loop through "*_displacement.png". Within the loop, that gives one of the three filenames. For the other two, use "set" with character substitution, changing "displacement" to "n" or "roughness".
